lib/HLSL/DxilValidation.cpp

```diff
@@ -534,70 +534,27 @@ struct ValidationContext {
     return I;
   }
 
-  void EmitInstrNote(Instruction *I, std::string Msg) {
-    Instruction *DbgI = GetDebugInstr(I);
-    const DebugLoc L = DbgI->getDebugLoc();
-    if (L) {
-      LastDebugLocEmit = L;
-    }
-
+  // Emit Error or note on instruction `I` with `Msg` for `Rule` if `isError` is true
+  void EmitInstrDiagMsg(Instruction *I, ValidationRule Rule, std::string Msg, bool isError = true) {
     BasicBlock *BB = I->getParent();
     Function *F = BB->getParent();
 
-    dxilutil::EmitErrorOnInstruction(DbgI, Msg);
-
-    // Add llvm information as a note to instruction string
-    std::string InstrStr;
-    raw_string_ostream InstrStream(InstrStr);
-    I->print(InstrStream, slotTracker);
-    InstrStream.flush();
-    StringRef InstrStrRef = InstrStr;
-    InstrStrRef = InstrStrRef.ltrim(); // Ignore indentation
-    Msg = "at '" + InstrStrRef.str() + "'";
-
-    // Print the parent block name
-    Msg += " in block '";
-    if (!BB->getName().empty()) {
-      Msg += BB->getName();
-    }
-    else {
-      unsigned idx = 0;
-      for (auto i = F->getBasicBlockList().begin(),
-        e = F->getBasicBlockList().end(); i != e; ++i) {
-        if (BB == &(*i)) {
-          break;
-        }
-        idx++;
-      }
-      Msg += "#" + std::to_string(idx);
-    }
-    Msg += "'";
-
-    // Print the function name
-    Msg += " of function '" + F->getName().str() + "'.";
-
-    dxilutil::EmitNoteOnContext(DbgI->getContext(), Msg);
-
-  }
-
-  void EmitInstrErrorMsg(Instruction *I, ValidationRule Rule, std::string Msg) {
     Instruction *DbgI = GetDebugInstr(I);
-    const DebugLoc L = DbgI->getDebugLoc();
-    if (L) {
-      // Instructions that get scalarized will likely hit
-      // this case. Avoid redundant diagnostic messages.
-      if (Rule == LastRuleEmit && L == LastDebugLocEmit) {
-        return;
```

Comment:
I think it's helpful to consider almost all of the above as removing EmitInstrNote outright and the below as the changes to EmitInstrErroMsg to add the ability to print a note as well as an error.

If I had it to do again, I'd have separated this into commits that alter EmitInstrErroMsg and then one that removes the  EmitInstrNote  function. I can still do that if you don't mind a force push.